### PR TITLE
Function to check rebalance status and time taken

### PR DESCRIPTION
**Pull request by Shrivaibavi**
This PR checks for
* Storage utilization , if its less then runs IO to increase storage utilization,
* Scale OSDs from 3-6, 
* Check for rebalance status 
* Reboot workers

Signed-off-by: Shrivaibavi Raghaventhiran <sraghave@redhat.com>

### CODE REVIEW

File: ocs_ci/ocs/cluster.py [hunk 1]
```diff
@@ -564,6 +571,42 @@ def get_throughput_percentage(self):
...
+            if states['state_name'] == 'active+clean':
+                if states['count'] != total_pg_count:
+                    logger.info("The rebalance is going on, "
+                                f"State of cluster is {ceph_health}")
+                    return True
+                return False
+
+    def time_taken_to_complete_rebalance(self, timeout=600):
+        """
+
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py]: Please add description

File: ocs_ci/ocs/cluster.py [hunk 2]
```diff
@@ -564,6 +571,42 @@ def get_throughput_percentage(self):
...
+        """
+        import time
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py]: Move to import section

File: ocs_ci/ocs/cluster.py [hunk 3]
```diff
@@ -564,6 +571,42 @@ def get_throughput_percentage(self):
...
+        flag = 1
+        while flag:
+            rc = self.get_rebalance_status()
+            if rc:
+                start_time = time.time()
```
Comment (Avilir) [inline ocs_ci/ocs/cluster.py]: This should be out of the loop, otherwize the end_time will be 0 all the time.

File: ocs_ci/ocs/cluster.py [hunk 4]
```diff
@@ -564,6 +571,42 @@ def get_throughput_percentage(self):
...
+                    return True
+                return False
+
+    def time_taken_to_complete_rebalance(self, timeout=600):
+        """
+
+        """
+        import time
+        flag = 1
+        while flag:
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py]: please use timeoutingSampler

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+from ocs_ci.ocs.resources import storage_cluster
+from tests import helpers
+from ocs_ci.utility.utils import ceph_health_check
+
+
+logger = logging.getLogger(__name__)
+
+
+@scale
+@tier4
```
Comment (RazTamir) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: please also add tier4a, b or c.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+@ignore_leftovers
```
Comment (RazTamir) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: why ignore leftovers?

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                sa_name = helpers.create_serviceaccount(
+                    pvc_obj.project.namespace
+                )
+
+                helpers.add_scc_policy(
+                    sa_name=sa_name.name, namespace=pvc_obj.project.namespace
+                )
+
+                pod_obj = helpers.create_pod(
+                    interface_type=constants.CEPHBLOCKPOOL,
```
Comment (RazTamir) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: what about cephfs?

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                # Create service_account to get privilege for deployment pods
+                sa_name = helpers.create_serviceaccount(
+                    pvc_obj.project.namespace
+                )
+
+                helpers.add_scc_policy(
+                    sa_name=sa_name.name, namespace=pvc_obj.project.namespace
+                )
+
+                pod_obj = helpers.create_pod(
```
Comment (fbalak) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: Maybe you can use dc_pod_factory which lets you specify `service_account`.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        osd_size = storage_cluster.get_osd_size()
+        count = storage_cluster.add_capacity(osd_size)
+        pod = OCP(
+            kind=constants.POD, namespace=config.ENV_DATA['cluster_namespace']
+        )
+        pod.wait_for_resource(
+            timeout=300,
+            condition=constants.STATUS_RUNNING,
+            selector='app=rook-ceph-osd',
+            resource_count=count * 3
```
Comment (fbalak) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: Why is there `count * 3`? What does `3` represent? Shouldn't it be loaded from somewhere dynamically?

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        )
+        assert ceph_health_check(), "Failed, New OSD pods failed to reach running state"
+
+        obj = CephCluster()
```
Comment (fbalak) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: I wouldn't name the variable `obj` but e.g. `cluster` but that's just a suggestion.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+        # Get rebalance status
+        rebalance_status = obj.get_rebalance_status()
+        if rebalance_status:
+            time_taken = obj.time_taken_to_complete_rebalance()
+            logger.info(time_taken)
+
+        worker_nodes = get_typed_nodes(node_type='worker')
+        nodes = AWSNodes()
```
Comment (ebenahar) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: is this test specific for AWS?
If not, please use PlatformNodesFactory and initiate it according to how it's done in https://github.com/red-hat-storage/ocs-ci/blob/75cc7c1b58f0d2d9c428bdfb3e9a204639c592c1/tests/conftest.py#L1465

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 6 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: We are scaling osds here, As we are not supporting cluster reduction for now, Added ignore_leftovers to ignore newly added OSDs.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 9 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: @fbalak 3 refers to number of OSDs... whereas count refers to 1 (ie 1 set of osd = 3 osds), The definition is also given in add_capacity, Please have a look at it.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 11 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: ACK, will address in next commit

File: ocs_ci/ocs/cluster.py [hunk 12]
```diff
@@ -564,6 +571,45 @@ def get_throughput_percentage(self):
...
+        """
+        This function calculates the time taken to complete
+        rebalance
+
+        Args:
+            timeout (int) : Timeout till when to check
+
+        """
+        start_time = time.time()
+        for sample in TimeoutSampler(
```
Comment (vasukulkarni) [inline ocs_ci/ocs/cluster.py]: s/sample/rebalance for readability

File: ocs_ci/ocs/cluster.py [hunk 13]
```diff
@@ -564,6 +571,45 @@ def get_throughput_percentage(self):
...
+            timeout=timeout, sleep=10, func=self.get_rebalance_status()
+        ):
+            while not sample:
```
Comment (vasukulkarni) [inline ocs_ci/ocs/cluster.py]: same as above

File: ocs_ci/ocs/cluster.py [hunk 14]
```diff
@@ -564,6 +571,45 @@ def get_throughput_percentage(self):
         logging.info(f"The throughput percentage of the cluster is {throughput_percentage}%")
         return throughput_percentage
 
+    def get_rebalance_status(self):
+        """
+        This function gets the rebalance status
+
+        Returns:
+            bool: True if rebalance is going on, False otherwise
```
Comment (vasukulkarni) [inline ocs_ci/ocs/cluster.py]: I think it makes sense to flip the Statu here, True if Rebalance is complete.

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 7 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: Added cephfs too

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 11 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: Addressed

File: ocs_ci/ocs/cluster.py (hunk 13 above)
Comment (Shrivaibavi) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py (hunk 14 above)
Comment (Shrivaibavi) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py (hunk 4 above)
Comment (Shrivaibavi) [reply ocs_ci/ocs/cluster.py]: done

File: ocs_ci/ocs/cluster.py [hunk 15]
```diff
@@ -586,6 +593,48 @@ def get_throughput_percentage(self):
         logging.info(f"The throughput percentage of the cluster is {throughput_percentage}%")
         return throughput_percentage
 
+    def get_rebalance_status(self):
+        """
+        This function gets the rebalance status
+
+        Returns:
+            bool: True if rebalance is completed, False otherwise
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:601]: Please add an empty line below

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+        # Get rebalance status
+        rebalance_status = cluster.get_rebalance_status()
+        if rebalance_status:
+            time_taken = cluster.time_taken_to_complete_rebalance()
+            logger.info(f"The time taken to complete rebalance {time_taken}")
+
+        worker_nodes = get_typed_nodes(node_type='worker')
+
+        node_obj = nodes()
```
Comment (ebenahar) [inline tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py:99]: nodes() is a fixture. The correct usage of it should be with injecting it into the test function. See reference in https://github.com/red-hat-storage/ocs-ci/blob/8dd733926df64676c136be3c76e62239443ac733/tests/manage/z_cluster/nodes/test_disk_failures.py#L143

File: ocs_ci/ocs/cluster.py (hunk 15 above)
Comment (Shrivaibavi) [reply ocs_ci/ocs/cluster.py:601]: DOne

File: tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py (hunk 16 above)
Comment (Shrivaibavi) [reply tests/e2e/scale/test_scale_osds_fill_75%_reboot_workers.py]: Done

File: ocs_ci/ocs/cluster.py [hunk 17]
```diff
@@ -586,6 +593,47 @@ def get_throughput_percentage(self):
...
+        """
+
+        ceph_status = self.get_ceph_status(format='json-pretty')
+        ceph_health = ceph_status['health']['status']
+        total_pg_count = ceph_status['pgmap']['num_pgs']
+        pg_states = ceph_status['pgmap']['pgs_by_state']
+        logger.info(ceph_health)
+        logger.info(pg_states)
+        for states in pg_states:
+            return (states['state_name'] == 'active+clean'
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py:612]: please fix indentation

File: ocs_ci/ocs/cluster.py [hunk 18]
```diff
@@ -586,6 +593,47 @@ def get_throughput_percentage(self):
...
+
+        """
+        start_time = time.time()
+        for rebalance in TimeoutSampler(
+            timeout=timeout, sleep=10, func=self.get_rebalance_status
+        ):
+            if rebalance:
+                logging.info("Rebalance is completed")
+                time_taken = time.time() - start_time
+                return (time_taken / 60)
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py:635]: nice!

File: ocs_ci/ocs/cluster.py (hunk 17 above)
Comment (Shrivaibavi) [reply ocs_ci/ocs/cluster.py]: DOne

File: ocs_ci/ocs/cluster.py [hunk 19]
```diff
@@ -471,14 +471,21 @@ def get_ceph_health(self, detail=False):
...
+            format (str) : Format of the output (e.g. json-pretty, json, plain)
+
         Returns:
             str: Output of the ceph status command.
 
         """
+        if format:
+            return self.toolbox.exec_cmd_on_pod(
+                f"ceph status -f {format}"
+            )
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:488]:
```suggestion
        cmd = "ceph status"
        if format:
            cmd += f" -f {format}"
        return self.toolbox.exec_cmd_on_pod(cmd)

```

File: ocs_ci/ocs/cluster.py [hunk 20]
```diff
@@ -471,14 +471,22 @@ def get_ceph_health(self, detail=False):
...
+
         Returns:
             str: Output of the ceph status command.
 
         """
+        cmd = "ceph status"
+        if format:
+            cmd += f" -f {format}"
+        return self.toolbox.exec_cmd_on_pod(cmd)
+
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:489]: lines 490-492 need to be removed

**Pull request merged**